# Python Class 8 — All Practice Questions Solved

## Complete Answer Key

This notebook contains **all 100 practice questions** from Python Class 8, solved in sequence.

For every question:
1. The **original question** is shown.
2. A **runnable solution** is provided.
3. The code is followed by a **line-by-line explanation**.
4. The notebook has been **executed**, so outputs are already included.

> **Note:** Questions that originally ask for user input use a fixed sample value so the notebook can run from top to bottom without waiting for manual input.


# Sets

## Question 1

**Question:** Create a set of five fruits, including at least one duplicate, and print its length.

In [1]:
fruits = {"apple", "banana", "apple", "mango", "orange"}
print("Set:", fruits)
print("Length:", len(fruits))

Set: {'banana', 'mango', 'apple', 'orange'}
Length: 4


### Line-by-line explanation

1. `fruits = {"apple", "banana", "apple", "mango", "orange"}` — creates a set; the duplicate `"apple"` is automatically kept only once.
2. `print("Set:", fruits)` — prints the resulting set. Set display order is not guaranteed.
3. `print("Length:", len(fruits))` — counts the unique values and prints the set length, which is 4.

## Question 2

**Question:** Create a set and add two new values to it using `.add()`.

In [2]:
numbers = {1, 2, 3}
numbers.add(4)
numbers.add(5)
print(numbers)

{1, 2, 3, 4, 5}


### Line-by-line explanation

1. `numbers = {1, 2, 3}` — creates the starting set.
2. `numbers.add(4)` — adds 4 to the set.
3. `numbers.add(5)` — adds 5 to the set.
4. `print(numbers)` — prints the updated set.

## Question 3

**Question:** Remove a value from a set using `.remove()`.

In [3]:
colors = {"red", "green", "blue"}
colors.remove("green")
print(colors)

{'blue', 'red'}


### Line-by-line explanation

1. `colors = {"red", "green", "blue"}` — creates a set containing three colors.
2. `colors.remove("green")` — removes `"green"` from the set.
3. `print(colors)` — prints the remaining values; their order is not guaranteed.

## Question 4

**Question:** Create two sets of numbers and print their union and intersection.

In [4]:
set_a = {1, 2, 3, 4}
set_b = {3, 4, 5, 6}

print("Union:", set_a | set_b)
print("Intersection:", set_a & set_b)

Union: {1, 2, 3, 4, 5, 6}
Intersection: {3, 4}


### Line-by-line explanation

1. `set_a = {1, 2, 3, 4}` — creates the first set.
2. `set_b = {3, 4, 5, 6}` — creates the second set.
3. `print("Union:", set_a | set_b)` — `|` combines all unique values from both sets.
4. `print("Intersection:", set_a & set_b)` — `&` keeps only values that appear in both sets.

## Question 5

**Question:** Explain, in your own words, one real situation where you would prefer a set over a list.

In [5]:
registered_students = {"Ali", "Sara", "Hamza", "Ali"}
print("Unique students:", registered_students)
print("Is Sara registered?", "Sara" in registered_students)

Unique students: {'Sara', 'Hamza', 'Ali'}
Is Sara registered? True


### Line-by-line explanation

1. `registered_students = {"Ali", "Sara", "Hamza", "Ali"}` — models a registration list where duplicate names should not be counted twice.
2. `print("Unique students:", registered_students)` — shows that the set stores each name only once.
3. `print("Is Sara registered?", "Sara" in registered_students)` — checks membership efficiently using the `in` operator.

# List Comprehensions

## Question 6

**Question:** Create a list comprehension that doubles every number in a list.

In [6]:
numbers = [1, 2, 3, 4, 5]
doubled = [number * 2 for number in numbers]
print(doubled)

[2, 4, 6, 8, 10]


### Line-by-line explanation

1. `numbers = [1, 2, 3, 4, 5]` — creates the original list.
2. `doubled = [number * 2 for number in numbers]` — loops through each number and places its doubled value into a new list.
3. `print(doubled)` — prints the new list.

## Question 7

**Question:** Create a list comprehension that keeps only numbers greater than 5.

In [7]:
numbers = [2, 5, 6, 8, 10, 3]
greater_than_five = [number for number in numbers if number > 5]
print(greater_than_five)

[6, 8, 10]


### Line-by-line explanation

1. `numbers = [2, 5, 6, 8, 10, 3]` — creates a list containing values on both sides of the condition.
2. `greater_than_five = [number for number in numbers if number > 5]` — keeps only numbers for which `number > 5` is true.
3. `print(greater_than_five)` — prints the filtered list.

## Question 8

**Question:** Create a list comprehension that converts a list of names to uppercase.

In [8]:
names = ["ali", "sara", "hamza"]
uppercase_names = [name.upper() for name in names]
print(uppercase_names)

['ALI', 'SARA', 'HAMZA']


### Line-by-line explanation

1. `names = ["ali", "sara", "hamza"]` — creates a list of lowercase names.
2. `uppercase_names = [name.upper() for name in names]` — calls `.upper()` on every name and stores the results.
3. `print(uppercase_names)` — prints the uppercase names.

## Question 9

**Question:** Rewrite one of your list comprehensions as a normal `for` loop and confirm it gives the same result.

In [9]:
numbers = [1, 2, 3, 4, 5]

comprehension_result = [number ** 2 for number in numbers]

loop_result = []
for number in numbers:
    loop_result.append(number ** 2)

print("Comprehension:", comprehension_result)
print("For loop:", loop_result)
print("Same result:", comprehension_result == loop_result)

Comprehension: [1, 4, 9, 16, 25]
For loop: [1, 4, 9, 16, 25]
Same result: True


### Line-by-line explanation

1. `numbers = [1, 2, 3, 4, 5]` — creates the input list.
2. `comprehension_result = [number ** 2 for number in numbers]` — creates squares using a list comprehension.
3. `loop_result = []` — creates an empty list for the normal loop.
4. `for number in numbers:` — visits each number one at a time.
5. `loop_result.append(number ** 2)` — calculates the square and adds it to the result list.
6. `print("Comprehension:", comprehension_result)` — prints the comprehension result.
7. `print("For loop:", loop_result)` — prints the loop result.
8. `print("Same result:", comprehension_result == loop_result)` — compares both lists and confirms they are equal.

## Question 10

**Question:** Create a list comprehension that keeps only words longer than 4 letters from a list of words.

In [10]:
words = ["cat", "apple", "banana", "book", "python"]
long_words = [word for word in words if len(word) > 4]
print(long_words)

['apple', 'banana', 'python']


### Line-by-line explanation

1. `words = ["cat", "apple", "banana", "book", "python"]` — creates a list of words with different lengths.
2. `long_words = [word for word in words if len(word) > 4]` — keeps words whose length is greater than 4.
3. `print(long_words)` — prints the filtered words.

# Exception Handling

## Question 11

**Question:** Write a `try`/`except` block that safely converts user input to an integer.

In [11]:
user_input = "25"

try:
    number = int(user_input)
    print("Integer:", number)
except ValueError:
    print("Invalid integer input.")

Integer: 25


### Line-by-line explanation

1. `user_input = "25"` — provides sample input so the notebook runs automatically.
2. `try:` — starts code that may raise an exception.
3. `number = int(user_input)` — attempts to convert the text into an integer.
4. `print("Integer:", number)` — prints the converted integer when conversion succeeds.
5. `except ValueError:` — handles invalid integer text without crashing the program.
6. `print("Invalid integer input.")` — shows a friendly error message if conversion fails.

## Question 12

**Question:** Add an `except ZeroDivisionError` to a calculator that divides two numbers.

In [12]:
a = 20
b = 0

try:
    result = a / b
    print("Result:", result)
except ZeroDivisionError:
    print("Cannot divide by zero.")

Cannot divide by zero.


### Line-by-line explanation

1. `a = 20` — stores the first number.
2. `b = 0` — stores zero as the denominator to demonstrate the exception.
3. `try:` — starts the potentially risky division.
4. `result = a / b` — attempts the division and raises `ZeroDivisionError` because the denominator is zero.
5. `print("Result:", result)` — would print the result if division succeeded.
6. `except ZeroDivisionError:` — catches the specific division-by-zero error.
7. `print("Cannot divide by zero.")` — prints a safe message instead of stopping the program.

## Question 13

**Question:** Add an `else` block that only runs when no error occurred.

In [13]:
a = 20
b = 5

try:
    result = a / b
except ZeroDivisionError:
    print("Cannot divide by zero.")
else:
    print("Result:", result)

Result: 4.0


### Line-by-line explanation

1. `a = 20` — stores the numerator.
2. `b = 5` — stores a non-zero denominator.
3. `try:` — starts the risky operation.
4. `result = a / b` — performs the division successfully.
5. `except ZeroDivisionError:` — would handle a zero denominator if one occurred.
6. `print("Cannot divide by zero.")` — would display the error message only when division fails.
7. `else:` — starts code that runs only when the `try` block succeeds.
8. `print("Result:", result)` — prints the successful result.

## Question 14

**Question:** Add a `finally` block that prints "Done" no matter what.

In [14]:
a = 10
b = 2

try:
    print("Result:", a / b)
except ZeroDivisionError:
    print("Cannot divide by zero.")
finally:
    print("Done")

Result: 5.0
Done


### Line-by-line explanation

1. `a = 10` — stores the numerator.
2. `b = 2` — stores a valid denominator.
3. `try:` — starts the operation that could fail.
4. `print("Result:", a / b)` — performs and prints the division.
5. `except ZeroDivisionError:` — handles a zero denominator if necessary.
6. `print("Cannot divide by zero.")` — prints the error message when division fails.
7. `finally:` — starts code that runs whether an exception occurred or not.
8. `print("Done")` — prints the completion message.

## Question 15

**Question:** Create a dictionary and use `try`/`except KeyError` to safely access a key that may not exist.

In [15]:
student = {"name": "Ali", "grade": "A"}

try:
    print(student["age"])
except KeyError:
    print("The 'age' key does not exist.")

The 'age' key does not exist.


### Line-by-line explanation

1. `student = {"name": "Ali", "grade": "A"}` — creates a dictionary without an `"age"` key.
2. `try:` — starts the potentially failing dictionary lookup.
3. `print(student["age"])` — attempts to access a missing key, which raises `KeyError`.
4. `except KeyError:` — catches the missing-key error.
5. `print("The 'age' key does not exist.")` — prints a clear message instead of crashing.

# File Handling

## Question 16

**Question:** Write three lines of text to a file called `diary.txt`.

In [16]:
with open("diary.txt", "w", encoding="utf-8") as file:
    file.write("Today I learned Python.\n")
    file.write("I practiced file handling.\n")
    file.write("I will keep learning.\n")

print("Three lines written to diary.txt")

Three lines written to diary.txt


### Line-by-line explanation

1. `with open("diary.txt", "w", encoding="utf-8") as file:` — opens or creates `diary.txt` in write mode and automatically closes it afterward.
2. `file.write("Today I learned Python.\n")` — writes the first line and `\n` moves to the next line.
3. `file.write("I practiced file handling.\n")` — writes the second line.
4. `file.write("I will keep learning.\n")` — writes the third line.
5. `print("Three lines written to diary.txt")` — confirms the operation completed.

## Question 17

**Question:** Read `diary.txt` back and print its contents.

In [17]:
with open("diary.txt", "r", encoding="utf-8") as file:
    content = file.read()

print(content)

Today I learned Python.
I practiced file handling.
I will keep learning.



### Line-by-line explanation

1. `with open("diary.txt", "r", encoding="utf-8") as file:` — opens the existing file in read mode.
2. `content = file.read()` — reads the complete file into `content`.
3. `print(content)` — prints all stored text.

## Question 18

**Question:** Use mode `"a"` to append one more line without erasing the existing ones.

In [18]:
with open("diary.txt", "a", encoding="utf-8") as file:
    file.write("I also practiced appending data.\n")

with open("diary.txt", "r", encoding="utf-8") as file:
    print(file.read())

Today I learned Python.
I practiced file handling.
I will keep learning.
I also practiced appending data.



### Line-by-line explanation

1. `with open("diary.txt", "a", encoding="utf-8") as file:` — opens the file in append mode, preserving existing content.
2. `file.write("I also practiced appending data.\n")` — adds a new line at the end.
3. `with open("diary.txt", "r", encoding="utf-8") as file:` — reopens the file for reading.
4. `print(file.read())` — reads and displays the complete file so the appended line can be verified.

## Question 19

**Question:** Use `.readlines()` instead of `.read()` and observe the difference in the result.

In [19]:
with open("diary.txt", "r", encoding="utf-8") as file:
    lines = file.readlines()

print(lines)
print("Type:", type(lines).__name__)

['Today I learned Python.\n', 'I practiced file handling.\n', 'I will keep learning.\n', 'I also practiced appending data.\n']
Type: list


### Line-by-line explanation

1. `with open("diary.txt", "r", encoding="utf-8") as file:` — opens the file for reading.
2. `lines = file.readlines()` — reads all lines into a list; each list item represents one line.
3. `print(lines)` — prints the list representation, including `\n` characters.
4. `print("Type:", type(lines).__name__)` — confirms that the returned value is a list.

## Question 20

**Question:** Explain in your own words why `with open(...) as file:` is safer than calling `open()` on its own.

In [20]:
file = open("diary.txt", "r", encoding="utf-8")
try:
    content = file.read()
    print(content)
finally:
    file.close()
    print("File closed:", file.closed)

Today I learned Python.
I practiced file handling.
I will keep learning.
I also practiced appending data.

File closed: True


### Line-by-line explanation

1. `file = open("diary.txt", "r", encoding="utf-8")` — opens the file manually.
2. `try:` — starts code that might fail while reading.
3. `content = file.read()` — reads the file.
4. `print(content)` — prints its contents.
5. `finally:` — ensures the cleanup code runs even if an error occurs.
6. `file.close()` — manually closes the file.
7. `print("File closed:", file.closed)` — verifies that the file is closed.

# Functions as First-Class Objects

## Question 21

**Question:** Write a small program based on this concept.

In [21]:
def greet(name):
    return f"Hello, {name}!"

message_function = greet
print(message_function("Saad"))

Hello, Saad!


### Line-by-line explanation

1. `def greet(name):` — defines a function named `greet` that accepts a name.
2. `return f"Hello, {name}!"` — returns a greeting string.
3. `message_function = greet` — stores a reference to the function in another variable; it does not call it yet.
4. `print(message_function("Saad"))` — calls the same function through the new variable and prints the returned message.

## Question 22

**Question:** Change the example so it accepts different input.

In [22]:
def greet(name):
    return f"Hello, {name}!"

message_function = greet
name = "Ayesha"
print(message_function(name))

Hello, Ayesha!


### Line-by-line explanation

1. `def greet(name):` — defines the reusable greeting function.
2. `return f"Hello, {name}!"` — builds and returns a greeting using the supplied name.
3. `message_function = greet` — assigns the function object to another variable.
4. `name = "Ayesha"` — provides a different input value.
5. `print(message_function(name))` — calls the function through the variable and prints the new greeting.

## Question 23

**Question:** Predict the output before running the code.

In [23]:
def greet(name):
    return f"Hello, {name}!"

message_function = greet
print(message_function("Saad"))

Hello, Saad!


### Line-by-line explanation

1. `def greet(name):` — defines the function.
2. `return f"Hello, {name}!"` — returns the greeting.
3. `message_function = greet` — makes `message_function` refer to `greet`.
4. `print(message_function("Saad"))` — calls the function and prints `Hello, Saad!`.

## Question 24

**Question:** Explain why the code works in your own words.

In [24]:
def greet(name):
    return f"Hello, {name}!"

message_function = greet
print(message_function("Saad"))

Hello, Saad!


### Line-by-line explanation

1. `def greet(name):` — creates the function object.
2. `return f"Hello, {name}!"` — returns a formatted greeting.
3. `message_function = greet` — shows that functions can be stored like other values.
4. `print(message_function("Saad"))` — uses the stored function reference to perform the call.

## Question 25

**Question:** Create one small real-world example using this concept.

In [25]:
def send_email(recipient):
    return f"Email prepared for {recipient}"

email_action = send_email
print(email_action("student@example.com"))

Email prepared for student@example.com


### Line-by-line explanation

1. `def send_email(recipient):` — defines a function representing an email-preparation action.
2. `return f"Email prepared for {recipient}"` — returns a message using the recipient.
3. `email_action = send_email` — stores the function in another variable, demonstrating first-class behavior.
4. `print(email_action("student@example.com"))` — calls the stored function reference.

# Passing a Function to Another Function

## Question 26

**Question:** Write a small program based on this concept.

In [26]:
def square(number):
    return number * number

def apply_operation(number, operation):
    return operation(number)

print(apply_operation(6, square))

36


### Line-by-line explanation

1. `def square(number):` — defines a function that calculates a square.
2. `return number * number` — returns the square.
3. `def apply_operation(number, operation):` — defines a function that accepts both a value and another function.
4. `return operation(number)` — calls the supplied function using the value.
5. `print(apply_operation(6, square))` — passes `square` as an argument and prints 36.

## Question 27

**Question:** Change the example so it accepts different input.

In [27]:
def square(number):
    return number * number

def apply_operation(number, operation):
    return operation(number)

value = 8
print(apply_operation(value, square))

64


### Line-by-line explanation

1. `def square(number):` — defines the square operation.
2. `return number * number` — returns the square.
3. `def apply_operation(number, operation):` — accepts a number and a function.
4. `return operation(number)` — executes the received function.
5. `value = 8` — changes the input value from the previous example.
6. `print(apply_operation(value, square))` — passes the value and function and prints 64.

## Question 28

**Question:** Predict the output before running the code.

In [28]:
def double(number):
    return number * 2

def apply_operation(number, operation):
    return operation(number)

print(apply_operation(5, double))

10


### Line-by-line explanation

1. `def double(number):` — defines a function that doubles a number.
2. `return number * 2` — returns twice the input.
3. `def apply_operation(number, operation):` — creates a function that accepts another function.
4. `return operation(number)` — calls the supplied operation.
5. `print(apply_operation(5, double))` — passes `double`, so the output is 10.

## Question 29

**Question:** Explain why the code works in your own words.

In [29]:
def cube(number):
    return number ** 3

def apply_operation(number, operation):
    return operation(number)

print(apply_operation(4, cube))

64


### Line-by-line explanation

1. `def cube(number):` — defines a cube operation.
2. `return number ** 3` — returns the cube of the number.
3. `def apply_operation(number, operation):` — receives a number and a function.
4. `return operation(number)` — runs the supplied function.
5. `print(apply_operation(4, cube))` — passes `cube` and prints 64.

## Question 30

**Question:** Create one small real-world example using this concept.

In [30]:
def apply_discount(price):
    return price * 0.90

def calculate(price, operation):
    return operation(price)

print(calculate(1000, apply_discount))

900.0


### Line-by-line explanation

1. `def apply_discount(price):` — defines a reusable discount operation.
2. `return price * 0.90` — calculates a 10% discount.
3. `def calculate(price, operation):` — accepts a price and a function describing what to do with it.
4. `return operation(price)` — executes the supplied pricing operation.
5. `print(calculate(1000, apply_discount))` — passes the discount function and prints 900.0.

# Returning a Function

## Question 31

**Question:** Write a small program based on this concept.

In [31]:
def create_multiplier(multiplier):
    def multiply(number):
        return number * multiplier
    return multiply

double = create_multiplier(2)
print(double(7))

14


### Line-by-line explanation

1. `def create_multiplier(multiplier):` — defines an outer function that receives a multiplier.
2. `def multiply(number):` — defines an inner function inside the outer function.
3. `return number * multiplier` — uses the inner function’s `number` and remembers the outer `multiplier`.
4. `return multiply` — returns the inner function itself.
5. `double = create_multiplier(2)` — creates a function that multiplies by 2.
6. `print(double(7))` — calls the returned function with 7 and prints 14.

## Question 32

**Question:** Change the example so it accepts different input.

In [32]:
def create_multiplier(multiplier):
    def multiply(number):
        return number * multiplier
    return multiply

triple = create_multiplier(3)
print(triple(5))

15


### Line-by-line explanation

1. `def create_multiplier(multiplier):` — defines the function factory.
2. `def multiply(number):` — creates the inner function.
3. `return number * multiplier` — uses the remembered multiplier.
4. `return multiply` — returns the inner function.
5. `triple = create_multiplier(3)` — creates a function that multiplies by 3.
6. `print(triple(5))` — uses the returned function with 5 and prints 15.

## Question 33

**Question:** Predict the output before running the code.

In [33]:
def create_multiplier(multiplier):
    def multiply(number):
        return number * multiplier
    return multiply

double = create_multiplier(2)
print(double(7))

14


### Line-by-line explanation

1. `def create_multiplier(multiplier):` — defines the outer function.
2. `def multiply(number):` — creates the inner function.
3. `return number * multiplier` — calculates using the captured multiplier.
4. `return multiply` — returns the inner function.
5. `double = create_multiplier(2)` — creates a multiplier-by-2 function.
6. `print(double(7))` — prints 14.

## Question 34

**Question:** Explain why the code works in your own words.

In [34]:
def create_multiplier(multiplier):
    def multiply(number):
        return number * multiplier
    return multiply

times_five = create_multiplier(5)
print(times_five(4))

20


### Line-by-line explanation

1. `def create_multiplier(multiplier):` — defines a function that builds multiplier functions.
2. `def multiply(number):` — defines the returned inner function.
3. `return number * multiplier` — performs multiplication using the outer value.
4. `return multiply` — returns the inner function.
5. `times_five = create_multiplier(5)` — creates a function that multiplies by 5.
6. `print(times_five(4))` — prints 20.

## Question 35

**Question:** Create one small real-world example using this concept.

In [35]:
def create_tax_calculator(rate):
    def calculate(price):
        return price * (1 + rate)
    return calculate

tax = create_tax_calculator(0.10)
print(tax(1000))

1100.0


### Line-by-line explanation

1. `def create_tax_calculator(rate):` — creates a function factory for a chosen tax rate.
2. `def calculate(price):` — defines the returned calculator.
3. `return price * (1 + rate)` — calculates the price including the remembered rate.
4. `return calculate` — returns the inner function.
5. `tax = create_tax_calculator(0.10)` — creates a calculator with a 10% rate.
6. `print(tax(1000))` — applies the calculator to 1000 and prints 1100.0.

# Lambda Functions

## Question 36

**Question:** Write a small program based on this concept.

In [36]:
square = lambda number: number * number
print(square(7))

49


### Line-by-line explanation

1. `square = lambda number: number * number` — creates a small anonymous function and stores it in `square`.
2. `print(square(7))` — calls the lambda with 7 and prints 49.

## Question 37

**Question:** Change the example so it accepts different input.

In [37]:
cube = lambda number: number ** 3
value = 4
print(cube(value))

64


### Line-by-line explanation

1. `cube = lambda number: number ** 3` — creates a lambda that calculates a cube.
2. `value = 4` — provides a different input.
3. `print(cube(value))` — calls the lambda and prints 64.

## Question 38

**Question:** Predict the output before running the code.

In [38]:
square = lambda number: number * number
print(square(7))

49


### Line-by-line explanation

1. `square = lambda number: number * number` — defines a one-expression function that returns the square.
2. `print(square(7))` — calls it with 7, so the predicted output is 49.

## Question 39

**Question:** Explain why the code works in your own words.

In [39]:
is_even = lambda number: number % 2 == 0
print(is_even(8))

True


### Line-by-line explanation

1. `is_even = lambda number: number % 2 == 0` — creates a lambda that returns a Boolean based on evenness.
2. `print(is_even(8))` — checks 8 and prints `True`.

## Question 40

**Question:** Create one small real-world example using this concept.

In [40]:
students = [("Ali", 85), ("Sara", 92), ("Hamza", 78)]
top_student = max(students, key=lambda student: student[1])
print(top_student)

('Sara', 92)


### Line-by-line explanation

1. `students = [("Ali", 85), ("Sara", 92), ("Hamza", 78)]` — stores student names with scores.
2. `top_student = max(students, key=lambda student: student[1])` — uses a lambda to tell `max()` to compare the second item, the score.
3. `print(top_student)` — prints the tuple containing the highest-scoring student.

# `map()` and `filter()` with Simple Functions

## Question 41

**Question:** Write a small program based on this concept.

In [41]:
numbers = [1, 2, 3, 4, 5]
squares = list(map(lambda number: number * number, numbers))
print(squares)

[1, 4, 9, 16, 25]


### Line-by-line explanation

1. `numbers = [1, 2, 3, 4, 5]` — creates the input list.
2. `squares = list(map(lambda number: number * number, numbers))` — applies the lambda to every item and converts the result to a list.
3. `print(squares)` — prints the squared values.

## Question 42

**Question:** Change the example so it accepts different input.

In [42]:
numbers = [3, 6, 9, 12, 15]
even_numbers = list(filter(lambda number: number % 2 == 0, numbers))
print(even_numbers)

[6, 12]


### Line-by-line explanation

1. `numbers = [3, 6, 9, 12, 15]` — provides different input values.
2. `even_numbers = list(filter(lambda number: number % 2 == 0, numbers))` — keeps only numbers for which the lambda returns `True`.
3. `print(even_numbers)` — prints the filtered even numbers.

## Question 43

**Question:** Predict the output before running the code.

In [43]:
numbers = [1, 2, 3, 4, 5]
squares = list(map(lambda number: number * number, numbers))
print(squares)

[1, 4, 9, 16, 25]


### Line-by-line explanation

1. `numbers = [1, 2, 3, 4, 5]` — contains the input values.
2. `squares = list(map(lambda number: number * number, numbers))` — squares every value using `map()`.
3. `print(squares)` — prints `[1, 4, 9, 16, 25]`.

## Question 44

**Question:** Explain why the code works in your own words.

In [44]:
numbers = [1, 2, 3, 4, 5]
even_numbers = list(filter(lambda number: number % 2 == 0, numbers))
print(even_numbers)

[2, 4]


### Line-by-line explanation

1. `numbers = [1, 2, 3, 4, 5]` — provides the input values.
2. `even_numbers = list(filter(lambda number: number % 2 == 0, numbers))` — keeps values whose remainder after division by 2 is zero.
3. `print(even_numbers)` — prints the even values `[2, 4]`.

## Question 45

**Question:** Create one small real-world example using this concept.

In [45]:
prices = [100, 250, 500, 1000]
discounted = list(map(lambda price: price * 0.9, prices))
expensive = list(filter(lambda price: price >= 500, prices))

print("Discounted:", discounted)
print("Expensive:", expensive)

Discounted: [90.0, 225.0, 450.0, 900.0]
Expensive: [500, 1000]


### Line-by-line explanation

1. `prices = [100, 250, 500, 1000]` — stores product prices.
2. `discounted = list(map(lambda price: price * 0.9, prices))` — applies a 10% reduction to every price.
3. `expensive = list(filter(lambda price: price >= 500, prices))` — keeps only prices of at least 500.
4. `print("Discounted:", discounted)` — prints the transformed prices.
5. `print("Expensive:", expensive)` — prints the filtered prices.

# Classes and Objects

## Question 46

**Question:** Write a small program based on this concept.

In [46]:
class Student:
    pass

student_1 = Student()
student_2 = Student()

print(type(student_1).__name__)
print(type(student_2).__name__)

Student
Student


### Line-by-line explanation

1. `class Student:` — defines a class named `Student`, acting as a blueprint.
2. `pass` — leaves the class body empty for this simple example.
3. `student_1 = Student()` — creates the first object from the class.
4. `student_2 = Student()` — creates a second independent object.
5. `print(type(student_1).__name__)` — shows that the first object is a `Student`.
6. `print(type(student_2).__name__)` — shows the same for the second object.

## Question 47

**Question:** Change the example so it accepts different input.

In [47]:
class Student:
    pass

student_1 = Student()
student_2 = Student()
student_3 = Student()

print(type(student_1).__name__)
print(type(student_2).__name__)
print(type(student_3).__name__)

Student
Student
Student


### Line-by-line explanation

1. `class Student:` — defines the blueprint.
2. `pass` — keeps the class body empty.
3. `student_1 = Student()` — creates one object.
4. `student_2 = Student()` — creates another object.
5. `student_3 = Student()` — creates a third object.
6. `print(type(student_1).__name__)` — prints `Student` for the first object.
7. `print(type(student_2).__name__)` — prints `Student` for the second object.
8. `print(type(student_3).__name__)` — prints `Student` for the third object.

## Question 48

**Question:** Predict the output before running the code.

In [48]:
class Student:
    pass

student_1 = Student()
student_2 = Student()

print(type(student_1).__name__)
print(type(student_2).__name__)

Student
Student


### Line-by-line explanation

1. `class Student:` — creates the class blueprint.
2. `pass` — means there are no custom attributes or methods yet.
3. `student_1 = Student()` — creates an instance.
4. `student_2 = Student()` — creates another instance.
5. `print(type(student_1).__name__)` — prints `Student`.
6. `print(type(student_2).__name__)` — prints `Student`.

## Question 49

**Question:** Explain why the code works in your own words.

In [49]:
class Student:
    pass

student = Student()
print(isinstance(student, Student))

True


### Line-by-line explanation

1. `class Student:` — defines the blueprint.
2. `pass` — keeps it empty.
3. `student = Student()` — creates an object from the class.
4. `print(isinstance(student, Student))` — checks whether the object belongs to the `Student` class and prints `True`.

## Question 50

**Question:** Create one small real-world example using this concept.

In [50]:
class Product:
    pass

laptop = Product()
phone = Product()

print("Laptop object type:", type(laptop).__name__)
print("Phone object type:", type(phone).__name__)

Laptop object type: Product
Phone object type: Product


### Line-by-line explanation

1. `class Product:` — defines a blueprint for product objects.
2. `pass` — keeps the class body empty for this basic example.
3. `laptop = Product()` — creates a product object representing a laptop.
4. `phone = Product()` — creates another product object representing a phone.
5. `print("Laptop object type:", type(laptop).__name__)` — confirms the laptop object is a `Product`.
6. `print("Phone object type:", type(phone).__name__)` — confirms the phone object is also a `Product`.

# The `__init__()` Constructor and `self`

## Question 51

**Question:** Write a small program based on this concept.

In [51]:
class Student:
    def __init__(self, name, age):
        self.name = name
        self.age = age

student = Student("Ali", 20)
print(student.name, student.age)

Ali 20


### Line-by-line explanation

1. `class Student:` — defines the class.
2. `def __init__(self, name, age):` — defines the constructor that runs when an object is created.
3. `self.name = name` — stores the name on the current object.
4. `self.age = age` — stores the age on the current object.
5. `student = Student("Ali", 20)` — creates an object and automatically calls `__init__`.
6. `print(student.name, student.age)` — reads and prints the object’s attributes.

## Question 52

**Question:** Change the example so it accepts different input.

In [52]:
class Student:
    def __init__(self, name, age):
        self.name = name
        self.age = age

student = Student("Sara", 19)
print(student.name, student.age)

Sara 19


### Line-by-line explanation

1. `class Student:` — defines the blueprint.
2. `def __init__(self, name, age):` — sets up the constructor parameters.
3. `self.name = name` — stores the provided name.
4. `self.age = age` — stores the provided age.
5. `student = Student("Sara", 19)` — uses different input values to create the object.
6. `print(student.name, student.age)` — prints Sara and 19.

## Question 53

**Question:** Predict the output before running the code.

In [53]:
class Student:
    def __init__(self, name, age):
        self.name = name
        self.age = age

student = Student("Ali", 20)
print(student.name)
print(student.age)

Ali
20


### Line-by-line explanation

1. `class Student:` — defines the class.
2. `def __init__(self, name, age):` — defines the constructor.
3. `self.name = name` — stores the name.
4. `self.age = age` — stores the age.
5. `student = Student("Ali", 20)` — creates the object with the supplied values.
6. `print(student.name)` — prints `Ali`.
7. `print(student.age)` — prints `20`.

## Question 54

**Question:** Explain why the code works in your own words.

In [54]:
class Student:
    def __init__(self, name, age):
        self.name = name
        self.age = age

student = Student("Ali", 20)
print(student.name, student.age)

Ali 20


### Line-by-line explanation

1. `class Student:` — creates the blueprint.
2. `def __init__(self, name, age):` — runs automatically during object creation.
3. `self.name = name` — creates an instance attribute for the name.
4. `self.age = age` — creates an instance attribute for the age.
5. `student = Student("Ali", 20)` — passes values into the constructor.
6. `print(student.name, student.age)` — prints the values stored on the object.

## Question 55

**Question:** Create one small real-world example using this concept.

In [55]:
class BankCustomer:
    def __init__(self, name, account_number):
        self.name = name
        self.account_number = account_number

customer = BankCustomer("Ali", "ACC-1001")
print(customer.name)
print(customer.account_number)

Ali
ACC-1001


### Line-by-line explanation

1. `class BankCustomer:` — defines a blueprint for customer objects.
2. `def __init__(self, name, account_number):` — defines the constructor parameters.
3. `self.name = name` — stores the customer name.
4. `self.account_number = account_number` — stores the account identifier on that object.
5. `customer = BankCustomer("Ali", "ACC-1001")` — creates a customer with real-world style data.
6. `print(customer.name)` — prints the customer name.
7. `print(customer.account_number)` — prints the account identifier.

# Instance Methods

## Question 56

**Question:** Write a small program based on this concept.

In [56]:
class Student:
    def __init__(self, name):
        self.name = name

    def greet(self):
        print(f"Hello, my name is {self.name}.")

student = Student("Ali")
student.greet()

Hello, my name is Ali.


### Line-by-line explanation

1. `class Student:` — defines the class.
2. `def __init__(self, name):` — creates the constructor.
3. `self.name = name` — stores the student’s name.
4. `def greet(self):` — defines an instance method.
5. `print(f"Hello, my name is {self.name}.")` — uses `self.name` to build the greeting.
6. `student = Student("Ali")` — creates a Student object.
7. `student.greet()` — calls the instance method through the object.

## Question 57

**Question:** Change the example so it accepts different input.

In [57]:
class Student:
    def __init__(self, name):
        self.name = name

    def greet(self):
        print(f"Hello, my name is {self.name}.")

student = Student("Sara")
student.greet()

Hello, my name is Sara.


### Line-by-line explanation

1. `class Student:` — defines the class.
2. `def __init__(self, name):` — defines the constructor.
3. `self.name = name` — stores the supplied name.
4. `def greet(self):` — defines a method that belongs to each Student object.
5. `print(f"Hello, my name is {self.name}.")` — prints the current object’s name.
6. `student = Student("Sara")` — uses a different input name.
7. `student.greet()` — calls the method for Sara.

## Question 58

**Question:** Predict the output before running the code.

In [58]:
class Student:
    def __init__(self, name):
        self.name = name

    def greet(self):
        print(f"Hello, my name is {self.name}.")

student = Student("Ali")
student.greet()

Hello, my name is Ali.


### Line-by-line explanation

1. `class Student:` — defines the class.
2. `def __init__(self, name):` — defines the constructor.
3. `self.name = name` — stores the object’s name.
4. `def greet(self):` — defines an instance method.
5. `print(f"Hello, my name is {self.name}.")` — uses the current object’s data.
6. `student = Student("Ali")` — creates a Student object.
7. `student.greet()` — calls the method, so the output is `Hello, my name is Ali.`.

## Question 59

**Question:** Explain why the code works in your own words.

In [59]:
class Student:
    def __init__(self, name):
        self.name = name

    def introduce(self):
        return f"My name is {self.name}."

student = Student("Hamza")
message = student.introduce()
print(message)

My name is Hamza.


### Line-by-line explanation

1. `class Student:` — defines the class.
2. `def __init__(self, name):` — defines how each object receives a name.
3. `self.name = name` — stores the name on the object.
4. `def introduce(self):` — defines a method that returns a sentence.
5. `return f"My name is {self.name}."` — returns a string using the object’s attribute.
6. `student = Student("Hamza")` — creates an object.
7. `message = student.introduce()` — calls the method and stores its return value.
8. `print(message)` — prints the returned message.

## Question 60

**Question:** Create one small real-world example using this concept.

In [60]:
class ShoppingCart:
    def __init__(self):
        self.total = 0

    def add_item(self, price):
        self.total += price

cart = ShoppingCart()
cart.add_item(250)
cart.add_item(150)
print("Total:", cart.total)

Total: 400


### Line-by-line explanation

1. `class ShoppingCart:` — defines a simple shopping-cart class.
2. `def __init__(self):` — defines the constructor with no extra input.
3. `self.total = 0` — starts the cart total at zero.
4. `def add_item(self, price):` — defines a method for adding an item price.
5. `self.total += price` — increases the cart total by the supplied price.
6. `cart = ShoppingCart()` — creates a cart object.
7. `cart.add_item(250)` — adds 250 to the total.
8. `cart.add_item(150)` — adds another 150.
9. `print("Total:", cart.total)` — prints the final total of 400.

# Updating Object Attributes

## Question 61

**Question:** Write a small program based on this concept.

In [61]:
class Student:
    def __init__(self, name, grade):
        self.name = name
        self.grade = grade

student = Student("Ali", "A")
print("Before:", student.grade)

student.grade = "A+"
print("After:", student.grade)

Before: A
After: A+


### Line-by-line explanation

1. `class Student:` — defines the class.
2. `def __init__(self, name, grade):` — defines the constructor.
3. `self.name = name` — stores the name.
4. `self.grade = grade` — stores the starting grade.
5. `student = Student("Ali", "A")` — creates the object with grade A.
6. `print("Before:", student.grade)` — prints the original grade.
7. `student.grade = "A+"` — updates the object’s grade attribute.
8. `print("After:", student.grade)` — prints the updated grade.

## Question 62

**Question:** Change the example so it accepts different input.

In [62]:
class Student:
    def __init__(self, name, grade):
        self.name = name
        self.grade = grade

student = Student("Sara", "B")
print("Before:", student.grade)

student.grade = "A"
print("After:", student.grade)

Before: B
After: A


### Line-by-line explanation

1. `class Student:` — defines the blueprint.
2. `def __init__(self, name, grade):` — defines the constructor.
3. `self.name = name` — stores the name.
4. `self.grade = grade` — stores the starting grade.
5. `student = Student("Sara", "B")` — creates the object using different input.
6. `print("Before:", student.grade)` — shows B before the change.
7. `student.grade = "A"` — updates the grade to A.
8. `print("After:", student.grade)` — shows the new grade.

## Question 63

**Question:** Predict the output before running the code.

In [63]:
class Student:
    def __init__(self, name, grade):
        self.name = name
        self.grade = grade

student = Student("Ali", "A")
student.grade = "A+"
print(student.grade)

A+


### Line-by-line explanation

1. `class Student:` — defines the class.
2. `def __init__(self, name, grade):` — defines the constructor.
3. `self.name = name` — stores the name.
4. `self.grade = grade` — stores the starting grade.
5. `student = Student("Ali", "A")` — creates the object.
6. `student.grade = "A+"` — changes the grade.
7. `print(student.grade)` — prints `A+`.

## Question 64

**Question:** Explain why the code works in your own words.

In [64]:
class Student:
    def __init__(self, name, grade):
        self.name = name
        self.grade = grade

student = Student("Ali", "A")
student.grade = "A+"
print(student.grade)

A+


### Line-by-line explanation

1. `class Student:` — defines the blueprint.
2. `def __init__(self, name, grade):` — sets up object initialization.
3. `self.name = name` — creates the name attribute.
4. `self.grade = grade` — creates the grade attribute.
5. `student = Student("Ali", "A")` — creates the object.
6. `student.grade = "A+"` — directly updates its state.
7. `print(student.grade)` — prints the updated state because attributes are mutable.

## Question 65

**Question:** Create one small real-world example using this concept.

In [65]:
class Light:
    def __init__(self):
        self.status = "Off"

    def turn_on(self):
        self.status = "On"

light = Light()
print("Before:", light.status)
light.turn_on()
print("After:", light.status)

Before: Off
After: On


### Line-by-line explanation

1. `class Light:` — defines a class representing a light.
2. `def __init__(self):` — defines the constructor.
3. `self.status = "Off"` — sets the initial state.
4. `def turn_on(self):` — defines a method that changes the state.
5. `self.status = "On"` — updates the attribute.
6. `light = Light()` — creates a light object.
7. `print("Before:", light.status)` — shows the initial state.
8. `light.turn_on()` — calls the method to update the state.
9. `print("After:", light.status)` — shows the updated state.

# Instance Variables vs Class Variables

## Question 66

**Question:** Write a small program based on this concept.

In [66]:
class Student:
    school_name = "Python Academy"

    def __init__(self, name):
        self.name = name

student_1 = Student("Ali")
student_2 = Student("Sara")

print(student_1.name, student_1.school_name)
print(student_2.name, student_2.school_name)

Ali Python Academy
Sara Python Academy


### Line-by-line explanation

1. `class Student:` — defines the Student class.
2. `school_name = "Python Academy"` — creates a class variable shared at class level.
3. `def __init__(self, name):` — defines the constructor.
4. `self.name = name` — creates an instance variable unique to each object.
5. `student_1 = Student("Ali")` — creates the first object.
6. `student_2 = Student("Sara")` — creates the second object with different instance data.
7. `print(student_1.name, student_1.school_name)` — shows Ali’s individual name and the shared school name.
8. `print(student_2.name, student_2.school_name)` — shows Sara’s individual name and the same shared school name.

## Question 67

**Question:** Change the example so it accepts different input.

In [67]:
class Student:
    school_name = "Python Academy"

    def __init__(self, name):
        self.name = name

student_1 = Student("Ali")
student_2 = Student("Sara")

Student.school_name = "Code Academy"

print(student_1.name, student_1.school_name)
print(student_2.name, student_2.school_name)

Ali Code Academy
Sara Code Academy


### Line-by-line explanation

1. `class Student:` — defines the class.
2. `school_name = "Python Academy"` — sets the original shared class variable.
3. `def __init__(self, name):` — defines the constructor.
4. `self.name = name` — creates a separate name for each object.
5. `student_1 = Student("Ali")` — creates the first student.
6. `student_2 = Student("Sara")` — creates the second student.
7. `Student.school_name = "Code Academy"` — changes the class variable for the class.
8. `print(student_1.name, student_1.school_name)` — shows that the first student sees the new shared value.
9. `print(student_2.name, student_2.school_name)` — shows the same shared value for the second student.

## Question 68

**Question:** Predict the output before running the code.

In [68]:
class Student:
    school_name = "Python Academy"

    def __init__(self, name):
        self.name = name

student_1 = Student("Ali")
student_2 = Student("Sara")

print(student_1.name)
print(student_1.school_name)

Ali
Python Academy


### Line-by-line explanation

1. `class Student:` — defines the class.
2. `school_name = "Python Academy"` — creates the shared class variable.
3. `def __init__(self, name):` — defines initialization.
4. `self.name = name` — creates an instance variable.
5. `student_1 = Student("Ali")` — creates an object with its own name.
6. `student_2 = Student("Sara")` — creates another object.
7. `print(student_1.name)` — prints the first object’s unique name.
8. `print(student_1.school_name)` — reads the class variable through the object.

## Question 69

**Question:** Explain why the code works in your own words.

In [69]:
class Student:
    school_name = "Python Academy"

    def __init__(self, name):
        self.name = name

student_1 = Student("Ali")
student_2 = Student("Sara")

print(student_1.name, student_2.name)
print(Student.school_name)

Ali Sara
Python Academy


### Line-by-line explanation

1. `class Student:` — defines the class.
2. `school_name = "Python Academy"` — stores a class-level value.
3. `def __init__(self, name):` — defines the constructor.
4. `self.name = name` — stores individual object data.
5. `student_1 = Student("Ali")` — creates the first object.
6. `student_2 = Student("Sara")` — creates the second object.
7. `print(student_1.name, student_2.name)` — shows that names differ between instances.
8. `print(Student.school_name)` — accesses the shared class variable directly through the class.

## Question 70

**Question:** Create one small real-world example using this concept.

In [70]:
class Employee:
    company = "TechWorks"

    def __init__(self, name):
        self.name = name

employee_1 = Employee("Ali")
employee_2 = Employee("Sara")

print(employee_1.name, employee_1.company)
print(employee_2.name, employee_2.company)

Ali TechWorks
Sara TechWorks


### Line-by-line explanation

1. `class Employee:` — defines an employee blueprint.
2. `company = "TechWorks"` — creates a shared class variable.
3. `def __init__(self, name):` — defines the constructor.
4. `self.name = name` — creates an instance variable unique to each employee.
5. `employee_1 = Employee("Ali")` — creates the first employee.
6. `employee_2 = Employee("Sara")` — creates the second employee.
7. `print(employee_1.name, employee_1.company)` — prints individual name plus shared company.
8. `print(employee_2.name, employee_2.company)` — does the same for the second employee.

# Class Methods

## Question 71

**Question:** Write a small program based on this concept.

In [71]:
class Student:
    school_name = "Python Academy"

    def __init__(self, name):
        self.name = name

    @classmethod
    def change_school(cls, new_name):
        cls.school_name = new_name

Student.change_school("Code Academy")
print(Student.school_name)

Code Academy


### Line-by-line explanation

1. `class Student:` — defines the class.
2. `school_name = "Python Academy"` — creates the class-level school name.
3. `def __init__(self, name):` — defines normal object initialization.
4. `self.name = name` — stores an individual student name.
5. `@classmethod` — marks the next method as a class method.
6. `def change_school(cls, new_name):` — defines a method whose first parameter `cls` refers to the class.
7. `cls.school_name = new_name` — changes the shared class variable.
8. `Student.change_school("Code Academy")` — calls the class method without creating an object.
9. `print(Student.school_name)` — prints the updated class value.

## Question 72

**Question:** Change the example so it accepts different input.

In [72]:
class Student:
    school_name = "Python Academy"

    @classmethod
    def change_school(cls, new_name):
        cls.school_name = new_name

Student.change_school("Future Academy")
print(Student.school_name)

Future Academy


### Line-by-line explanation

1. `class Student:` — defines the class.
2. `school_name = "Python Academy"` — sets the initial class variable.
3. `@classmethod` — turns `change_school` into a class method.
4. `def change_school(cls, new_name):` — receives the class through `cls` and a new school name.
5. `cls.school_name = new_name` — updates the shared class variable.
6. `Student.change_school("Future Academy")` — uses a different input value to update the class.
7. `print(Student.school_name)` — prints the new shared value.

## Question 73

**Question:** Predict the output before running the code.

In [73]:
class Student:
    school_name = "Python Academy"

    @classmethod
    def change_school(cls, new_name):
        cls.school_name = new_name

Student.change_school("Code Academy")
print(Student.school_name)

Code Academy


### Line-by-line explanation

1. `class Student:` — creates the class.
2. `school_name = "Python Academy"` — stores the initial shared value.
3. `@classmethod` — makes the following method operate on the class.
4. `def change_school(cls, new_name):` — defines the class method.
5. `cls.school_name = new_name` — changes the class variable.
6. `Student.change_school("Code Academy")` — calls the method and changes the school name.
7. `print(Student.school_name)` — prints `Code Academy`.

## Question 74

**Question:** Explain why the code works in your own words.

In [74]:
class Student:
    school_name = "Python Academy"

    @classmethod
    def change_school(cls, new_name):
        cls.school_name = new_name

Student.change_school("Code Academy")
print(Student.school_name)

Code Academy


### Line-by-line explanation

1. `class Student:` — defines the class blueprint.
2. `school_name = "Python Academy"` — creates a class variable.
3. `@classmethod` — tells Python the method receives the class as `cls`.
4. `def change_school(cls, new_name):` — defines the method that changes shared class state.
5. `cls.school_name = new_name` — updates the class variable.
6. `Student.change_school("Code Academy")` — calls the method on the class itself.
7. `print(Student.school_name)` — prints the updated value.

## Question 75

**Question:** Create one small real-world example using this concept.

In [75]:
class Library:
    total_books = 100

    @classmethod
    def add_books(cls, amount):
        cls.total_books += amount

Library.add_books(25)
print("Total books:", Library.total_books)

Total books: 125


### Line-by-line explanation

1. `class Library:` — defines a library class.
2. `total_books = 100` — stores a shared class-level count.
3. `@classmethod` — makes `add_books` a class method.
4. `def add_books(cls, amount):` — receives the class and number of books to add.
5. `cls.total_books += amount` — updates the shared count.
6. `Library.add_books(25)` — adds 25 books using the class method.
7. `print("Total books:", Library.total_books)` — prints the new total of 125.

# Encapsulation

## Question 76

**Question:** Write a small program based on this concept.

In [76]:
class BankAccount:
    def __init__(self, balance):
        self.__balance = balance

    def deposit(self, amount):
        if amount > 0:
            self.__balance += amount

    def get_balance(self):
        return self.__balance

account = BankAccount(1000)
account.deposit(500)
print("Balance:", account.get_balance())

Balance: 1500


### Line-by-line explanation

1. `class BankAccount:` — defines the account class.
2. `def __init__(self, balance):` — defines the constructor.
3. `self.__balance = balance` — stores balance in a double-underscore attribute, which Python name-mangles.
4. `def deposit(self, amount):` — defines controlled access for adding money.
5. `if amount > 0:` — validates that the deposit is positive.
6. `self.__balance += amount` — updates the protected balance internally.
7. `def get_balance(self):` — defines a method for reading the balance.
8. `return self.__balance` — returns the internal balance.
9. `account = BankAccount(1000)` — creates an account with an initial balance.
10. `account.deposit(500)` — changes the balance through the controlled method.
11. `print("Balance:", account.get_balance())` — reads the balance through the public method.

## Question 77

**Question:** Change the example so it accepts different input.

In [77]:
class BankAccount:
    def __init__(self, balance):
        self.__balance = balance

    def deposit(self, amount):
        if amount > 0:
            self.__balance += amount

    def get_balance(self):
        return self.__balance

account = BankAccount(2000)
account.deposit(250)
print("Balance:", account.get_balance())

Balance: 2250


### Line-by-line explanation

1. `class BankAccount:` — defines the class.
2. `def __init__(self, balance):` — initializes the account.
3. `self.__balance = balance` — stores the starting balance internally.
4. `def deposit(self, amount):` — defines a controlled deposit operation.
5. `if amount > 0:` — rejects non-positive deposits.
6. `self.__balance += amount` — adds the valid amount.
7. `def get_balance(self):` — defines a safe read method.
8. `return self.__balance` — returns the current balance.
9. `account = BankAccount(2000)` — uses a different starting balance.
10. `account.deposit(250)` — adds 250 through the method.
11. `print("Balance:", account.get_balance())` — prints 2250.

## Question 78

**Question:** Predict the output before running the code.

In [78]:
class BankAccount:
    def __init__(self, balance):
        self.__balance = balance

    def deposit(self, amount):
        if amount > 0:
            self.__balance += amount

    def get_balance(self):
        return self.__balance

account = BankAccount(1000)
account.deposit(500)
print(account.get_balance())

1500


### Line-by-line explanation

1. `class BankAccount:` — defines the class.
2. `def __init__(self, balance):` — sets up the object.
3. `self.__balance = balance` — stores the balance internally.
4. `def deposit(self, amount):` — defines the controlled update method.
5. `if amount > 0:` — checks that the deposit is valid.
6. `self.__balance += amount` — updates the internal value.
7. `def get_balance(self):` — provides controlled access to the value.
8. `return self.__balance` — returns the current balance.
9. `account = BankAccount(1000)` — creates the account.
10. `account.deposit(500)` — raises the balance to 1500.
11. `print(account.get_balance())` — prints 1500.

## Question 79

**Question:** Explain why the code works in your own words.

In [79]:
class BankAccount:
    def __init__(self, balance):
        self.__balance = balance

    def deposit(self, amount):
        if amount > 0:
            self.__balance += amount

    def get_balance(self):
        return self.__balance

account = BankAccount(1000)
account.deposit(500)
print(account.get_balance())

1500


### Line-by-line explanation

1. `class BankAccount:` — defines the blueprint.
2. `def __init__(self, balance):` — initializes each account.
3. `self.__balance = balance` — keeps the balance as an internal attribute.
4. `def deposit(self, amount):` — defines the public operation for deposits.
5. `if amount > 0:` — validates the amount.
6. `self.__balance += amount` — changes the balance only after validation.
7. `def get_balance(self):` — defines a method for reading the balance.
8. `return self.__balance` — returns the internal value.
9. `account = BankAccount(1000)` — creates the account.
10. `account.deposit(500)` — uses the public method to change state.
11. `print(account.get_balance())` — reads the result through the public method.

## Question 80

**Question:** Create one small real-world example using this concept.

In [80]:
class BankAccount:
    def __init__(self, balance):
        self.__balance = balance

    def deposit(self, amount):
        if amount > 0:
            self.__balance += amount

    def get_balance(self):
        return self.__balance

account = BankAccount(1000)
account.deposit(250)
print("Account balance:", account.get_balance())

Account balance: 1250


### Line-by-line explanation

1. `class BankAccount:` — models a bank account.
2. `def __init__(self, balance):` — sets the starting balance.
3. `self.__balance = balance` — keeps the balance internal.
4. `def deposit(self, amount):` — provides a controlled deposit operation.
5. `if amount > 0:` — validates the deposit.
6. `self.__balance += amount` — updates the balance when valid.
7. `def get_balance(self):` — provides controlled read access.
8. `return self.__balance` — returns the stored balance.
9. `account = BankAccount(1000)` — creates the account.
10. `account.deposit(250)` — adds a valid amount.
11. `print("Account balance:", account.get_balance())` — prints 1250.

# Inheritance

## Question 81

**Question:** Write a small program based on this concept.

In [81]:
class Person:
    def __init__(self, name):
        self.name = name

    def introduce(self):
        print(f"My name is {self.name}.")

class Student(Person):
    pass

student = Student("Ali")
student.introduce()

My name is Ali.


### Line-by-line explanation

1. `class Person:` — defines the parent class.
2. `def __init__(self, name):` — defines the parent constructor.
3. `self.name = name` — stores the name.
4. `def introduce(self):` — defines behavior shared with child classes.
5. `print(f"My name is {self.name}.")` — prints the introduction.
6. `class Student(Person):` — creates a child class that inherits from `Person`.
7. `pass` — adds no new behavior, so the parent behavior is reused.
8. `student = Student("Ali")` — creates a Student using the inherited constructor.
9. `student.introduce()` — calls the inherited method.

## Question 82

**Question:** Change the example so it accepts different input.

In [82]:
class Person:
    def __init__(self, name):
        self.name = name

    def introduce(self):
        print(f"My name is {self.name}.")

class Teacher(Person):
    pass

teacher = Teacher("Sara")
teacher.introduce()

My name is Sara.


### Line-by-line explanation

1. `class Person:` — defines the parent class.
2. `def __init__(self, name):` — creates the reusable constructor.
3. `self.name = name` — stores the name.
4. `def introduce(self):` — defines inherited behavior.
5. `print(f"My name is {self.name}.")` — prints the name.
6. `class Teacher(Person):` — creates a child class inheriting from Person.
7. `pass` — uses the parent implementation unchanged.
8. `teacher = Teacher("Sara")` — creates a Teacher object using the inherited constructor.
9. `teacher.introduce()` — uses the inherited method.

## Question 83

**Question:** Predict the output before running the code.

In [83]:
class Person:
    def __init__(self, name):
        self.name = name

    def introduce(self):
        print(f"My name is {self.name}.")

class Student(Person):
    pass

student = Student("Ali")
student.introduce()

My name is Ali.


### Line-by-line explanation

1. `class Person:` — defines the parent class.
2. `def __init__(self, name):` — defines the inherited constructor.
3. `self.name = name` — stores the name.
4. `def introduce(self):` — defines the inherited method.
5. `print(f"My name is {self.name}.")` — prints the introduction.
6. `class Student(Person):` — makes Student inherit from Person.
7. `pass` — means Student adds nothing new.
8. `student = Student("Ali")` — creates a Student object.
9. `student.introduce()` — uses the inherited method, producing `My name is Ali.`.

## Question 84

**Question:** Explain why the code works in your own words.

In [84]:
class Vehicle:
    def move(self):
        print("Vehicle is moving")

class Car(Vehicle):
    pass

car = Car()
car.move()

Vehicle is moving


### Line-by-line explanation

1. `class Vehicle:` — defines the parent class.
2. `def move(self):` — defines behavior available to children.
3. `print("Vehicle is moving")` — prints the movement message.
4. `class Car(Vehicle):` — creates a child class inheriting from Vehicle.
5. `pass` — keeps the inherited method unchanged.
6. `car = Car()` — creates a Car object.
7. `car.move()` — calls the inherited `move()` method.

## Question 85

**Question:** Create one small real-world example using this concept.

In [85]:
class Employee:
    def __init__(self, name):
        self.name = name

    def work(self):
        print(f"{self.name} is working.")

class Developer(Employee):
    pass

developer = Developer("Ali")
developer.work()

Ali is working.


### Line-by-line explanation

1. `class Employee:` — defines a reusable parent class.
2. `def __init__(self, name):` — defines the parent constructor.
3. `self.name = name` — stores the employee name.
4. `def work(self):` — defines common employee behavior.
5. `print(f"{self.name} is working.")` — prints a work message.
6. `class Developer(Employee):` — creates a child class inheriting from Employee.
7. `pass` — reuses the parent behavior unchanged.
8. `developer = Developer("Ali")` — creates a Developer object.
9. `developer.work()` — calls the inherited method.

# Method Overriding and Polymorphism

## Question 86

**Question:** Write a small program based on this concept.

In [86]:
class Animal:
    def sound(self):
        print("Some animal sound")

class Dog(Animal):
    def sound(self):
        print("Dog sound")

animal = Animal()
dog = Dog()

animal.sound()
dog.sound()

Some animal sound
Dog sound


### Line-by-line explanation

1. `class Animal:` — defines a general parent class.
2. `def sound(self):` — defines a general `sound()` method.
3. `print("Some animal sound")` — provides the parent implementation.
4. `class Dog(Animal):` — creates a child class.
5. `def sound(self):` — overrides the parent’s `sound()` method.
6. `print("Dog sound")` — provides Dog’s own behavior.
7. `animal = Animal()` — creates a parent-class object.
8. `dog = Dog()` — creates a Dog object.
9. `animal.sound()` — uses the parent implementation.
10. `dog.sound()` — uses the overridden Dog implementation.

## Question 87

**Question:** Change the example so it accepts different input.

In [87]:
class Animal:
    def sound(self):
        print("Some animal sound")

class Cat(Animal):
    def sound(self):
        print("Cat sound")

animal = Animal()
cat = Cat()

animal.sound()
cat.sound()

Some animal sound
Cat sound


### Line-by-line explanation

1. `class Animal:` — defines the parent class.
2. `def sound(self):` — defines the general method.
3. `print("Some animal sound")` — prints the generic sound.
4. `class Cat(Animal):` — creates a child class.
5. `def sound(self):` — overrides the inherited method.
6. `print("Cat sound")` — defines Cat’s behavior.
7. `animal = Animal()` — creates an Animal object.
8. `cat = Cat()` — creates a Cat object.
9. `animal.sound()` — calls the parent version.
10. `cat.sound()` — calls the overridden Cat version.

## Question 88

**Question:** Predict the output before running the code.

In [88]:
class Animal:
    def sound(self):
        print("Some animal sound")

class Dog(Animal):
    def sound(self):
        print("Dog sound")

animal = Animal()
dog = Dog()

animal.sound()
dog.sound()

Some animal sound
Dog sound


### Line-by-line explanation

1. `class Animal:` — defines the parent class.
2. `def sound(self):` — creates the general sound method.
3. `print("Some animal sound")` — is the parent behavior.
4. `class Dog(Animal):` — creates the child class.
5. `def sound(self):` — overrides the same method name.
6. `print("Dog sound")` — defines Dog-specific behavior.
7. `animal = Animal()` — creates the parent object.
8. `dog = Dog()` — creates the child object.
9. `animal.sound()` — prints `Some animal sound`.
10. `dog.sound()` — prints `Dog sound`; the same method name behaves differently.

## Question 89

**Question:** Explain why the code works in your own words.

In [89]:
class Animal:
    def sound(self):
        print("Some animal sound")

class Dog(Animal):
    def sound(self):
        print("Dog sound")

def make_sound(animal):
    animal.sound()

make_sound(Animal())
make_sound(Dog())

Some animal sound
Dog sound


### Line-by-line explanation

1. `class Animal:` — defines the general parent class.
2. `def sound(self):` — defines the common method interface.
3. `print("Some animal sound")` — provides the default behavior.
4. `class Dog(Animal):` — creates a child class.
5. `def sound(self):` — overrides the common method.
6. `print("Dog sound")` — provides Dog-specific behavior.
7. `def make_sound(animal):` — defines code that accepts any compatible object.
8. `animal.sound()` — calls `sound()` without needing to know the exact class.
9. `make_sound(Animal())` — uses the Animal implementation.
10. `make_sound(Dog())` — uses the Dog implementation, demonstrating polymorphism.

## Question 90

**Question:** Create one small real-world example using this concept.

In [90]:
class Payment:
    def pay(self):
        print("Processing payment")

class CardPayment(Payment):
    def pay(self):
        print("Paid by card")

class CashPayment(Payment):
    def pay(self):
        print("Paid by cash")

payments = [CardPayment(), CashPayment()]

for payment in payments:
    payment.pay()

Paid by card
Paid by cash


### Line-by-line explanation

1. `class Payment:` — defines a common payment interface.
2. `def pay(self):` — defines the common method name.
3. `print("Processing payment")` — provides a general implementation.
4. `class CardPayment(Payment):` — creates a payment type that inherits from Payment.
5. `def pay(self):` — overrides the payment behavior for cards.
6. `print("Paid by card")` — prints the card-specific behavior.
7. `class CashPayment(Payment):` — creates another payment type.
8. `def pay(self):` — overrides the same method for cash.
9. `print("Paid by cash")` — prints cash-specific behavior.
10. `payments = [CardPayment(), CashPayment()]` — stores different objects in one list.
11. `for payment in payments:` — loops through both objects.
12. `payment.pay()` — calls the same method name, but each object responds in its own way.

# Abstraction with Abstract Base Classes

## Question 91

**Question:** Write a small program based on this concept.

In [91]:
from abc import ABC, abstractmethod

class Shape(ABC):
    @abstractmethod
    def area(self):
        pass

class Square(Shape):
    def __init__(self, side):
        self.side = side

    def area(self):
        return self.side ** 2

square = Square(5)
print("Area:", square.area())

Area: 25


### Line-by-line explanation

1. `from abc import ABC, abstractmethod` — imports the tools for abstract base classes.
2. `class Shape(ABC):` — defines an abstract base class.
3. `@abstractmethod` — marks `area()` as required for child classes.
4. `def area(self):` — declares the required method.
5. `pass` — leaves the abstract method without a concrete implementation.
6. `class Square(Shape):` — creates a concrete child class.
7. `def __init__(self, side):` — defines the square constructor.
8. `self.side = side` — stores the side length.
9. `def area(self):` — implements the required abstract method.
10. `return self.side ** 2` — calculates the square’s area.
11. `square = Square(5)` — creates a concrete Square object.
12. `print("Area:", square.area())` — calls the implemented method and prints 25.

## Question 92

**Question:** Change the example so it accepts different input.

In [92]:
from abc import ABC, abstractmethod

class Shape(ABC):
    @abstractmethod
    def area(self):
        pass

class Circle(Shape):
    def __init__(self, radius):
        self.radius = radius

    def area(self):
        return 3.14 * self.radius ** 2

circle = Circle(3)
print("Area:", circle.area())

Area: 28.26


### Line-by-line explanation

1. `from abc import ABC, abstractmethod` — imports the abstract-class tools.
2. `class Shape(ABC):` — creates the abstract base class.
3. `@abstractmethod` — requires child classes to implement `area()`.
4. `def area(self):` — declares the required interface.
5. `pass` — leaves the abstract method unimplemented here.
6. `class Circle(Shape):` — creates a concrete child class.
7. `def __init__(self, radius):` — defines the circle constructor.
8. `self.radius = radius` — stores the radius.
9. `def area(self):` — implements the required method.
10. `return 3.14 * self.radius ** 2` — calculates the circle area.
11. `circle = Circle(3)` — creates a concrete Circle object.
12. `print("Area:", circle.area())` — prints the calculated area, 28.26.

## Question 93

**Question:** Predict the output before running the code.

In [93]:
from abc import ABC, abstractmethod

class Shape(ABC):
    @abstractmethod
    def area(self):
        pass

class Square(Shape):
    def __init__(self, side):
        self.side = side

    def area(self):
        return self.side ** 2

square = Square(5)
print("Area:", square.area())

Area: 25


### Line-by-line explanation

1. `from abc import ABC, abstractmethod` — imports ABC and the decorator.
2. `class Shape(ABC):` — creates an abstract base class.
3. `@abstractmethod` — marks `area()` as mandatory.
4. `def area(self):` — declares the required operation.
5. `pass` — does not provide a concrete implementation.
6. `class Square(Shape):` — creates a concrete implementation.
7. `def __init__(self, side):` — initializes the square.
8. `self.side = side` — stores the side.
9. `def area(self):` — implements the abstract method.
10. `return self.side ** 2` — calculates the area.
11. `square = Square(5)` — creates the object.
12. `print("Area:", square.area())` — prints 25.

## Question 94

**Question:** Explain why the code works in your own words.

In [94]:
from abc import ABC, abstractmethod

class Vehicle(ABC):
    @abstractmethod
    def start(self):
        pass

class Car(Vehicle):
    def start(self):
        return "Car started"

car = Car()
print(car.start())

Car started


### Line-by-line explanation

1. `from abc import ABC, abstractmethod` — imports abstract-class support.
2. `class Vehicle(ABC):` — defines an abstract vehicle class.
3. `@abstractmethod` — marks `start()` as required.
4. `def start(self):` — declares the required interface.
5. `pass` — leaves implementation to child classes.
6. `class Car(Vehicle):` — creates a concrete child class.
7. `def start(self):` — implements the required method.
8. `return "Car started"` — returns the concrete behavior.
9. `car = Car()` — creates a valid concrete object.
10. `print(car.start())` — calls the implemented method and prints the result.

## Question 95

**Question:** Create one small real-world example using this concept.

In [95]:
from abc import ABC, abstractmethod

class Report(ABC):
    @abstractmethod
    def generate(self):
        pass

class SalesReport(Report):
    def generate(self):
        return "Sales report generated"

report = SalesReport()
print(report.generate())

Sales report generated


### Line-by-line explanation

1. `from abc import ABC, abstractmethod` — imports the abstract base class tools.
2. `class Report(ABC):` — defines an abstract report interface.
3. `@abstractmethod` — requires concrete report classes to implement `generate()`.
4. `def generate(self):` — declares the required method.
5. `pass` — leaves implementation to subclasses.
6. `class SalesReport(Report):` — creates a concrete report type.
7. `def generate(self):` — implements the required method.
8. `return "Sales report generated"` — returns the concrete report message.
9. `report = SalesReport()` — creates a concrete report object.
10. `print(report.generate())` — calls the implemented method.

# Building a Small OOP Program

## Question 96

**Question:** Write a small program based on this concept.

In [96]:
class Product:
    def __init__(self, name, price):
        self.name = name
        self.price = price

    def apply_discount(self, percent):
        if 0 <= percent <= 100:
            self.price -= self.price * percent / 100

product = Product("Laptop", 100000)
product.apply_discount(10)

print("Product:", product.name)
print("Final price:", product.price)

Product: Laptop
Final price: 90000.0


### Line-by-line explanation

1. `class Product:` — defines the product blueprint.
2. `def __init__(self, name, price):` — defines the constructor.
3. `self.name = name` — stores the product name.
4. `self.price = price` — stores the product price.
5. `def apply_discount(self, percent):` — defines a method for changing the price.
6. `if 0 <= percent <= 100:` — validates that the discount percentage is sensible.
7. `self.price -= self.price * percent / 100` — calculates and subtracts the discount.
8. `product = Product("Laptop", 100000)` — creates a product object.
9. `product.apply_discount(10)` — applies a 10% discount.
10. `print("Product:", product.name)` — prints the product name.
11. `print("Final price:", product.price)` — prints the discounted price of 90000.0.

## Question 97

**Question:** Change the example so it accepts different input.

In [97]:
class Product:
    def __init__(self, name, price):
        self.name = name
        self.price = price

    def apply_discount(self, percent):
        if 0 <= percent <= 100:
            self.price -= self.price * percent / 100

product = Product("Phone", 60000)
product.apply_discount(15)

print("Product:", product.name)
print("Final price:", product.price)

Product: Phone
Final price: 51000.0


### Line-by-line explanation

1. `class Product:` — defines the product class.
2. `def __init__(self, name, price):` — sets up product data.
3. `self.name = name` — stores the product name.
4. `self.price = price` — stores the starting price.
5. `def apply_discount(self, percent):` — defines the discount operation.
6. `if 0 <= percent <= 100:` — checks that the percentage is valid.
7. `self.price -= self.price * percent / 100` — reduces the price by the selected percentage.
8. `product = Product("Phone", 60000)` — uses different product input.
9. `product.apply_discount(15)` — applies a 15% discount.
10. `print("Product:", product.name)` — prints the product name.
11. `print("Final price:", product.price)` — prints 51000.0.

## Question 98

**Question:** Predict the output before running the code.

In [98]:
class Product:
    def __init__(self, name, price):
        self.name = name
        self.price = price

    def apply_discount(self, percent):
        if 0 <= percent <= 100:
            self.price -= self.price * percent / 100

product = Product("Laptop", 100000)
product.apply_discount(10)

print("Product:", product.name)
print("Final price:", product.price)

Product: Laptop
Final price: 90000.0


### Line-by-line explanation

1. `class Product:` — defines the product class.
2. `def __init__(self, name, price):` — defines how a product is initialized.
3. `self.name = name` — stores the name.
4. `self.price = price` — stores the original price.
5. `def apply_discount(self, percent):` — defines the discount method.
6. `if 0 <= percent <= 100:` — checks the discount range.
7. `self.price -= self.price * percent / 100` — subtracts 10% of the original price, leaving 90000.0.
8. `product = Product("Laptop", 100000)` — creates the product.
9. `product.apply_discount(10)` — changes the product’s price.
10. `print("Product:", product.name)` — prints `Laptop`.
11. `print("Final price:", product.price)` — prints `90000.0`.

## Question 99

**Question:** Explain why the code works in your own words.

In [99]:
class Product:
    def __init__(self, name, price):
        self.name = name
        self.price = price

    def apply_discount(self, percent):
        if 0 <= percent <= 100:
            self.price -= self.price * percent / 100

product = Product("Laptop", 100000)
product.apply_discount(20)
print(product.price)

80000.0


### Line-by-line explanation

1. `class Product:` — defines the blueprint.
2. `def __init__(self, name, price):` — defines the constructor.
3. `self.name = name` — stores the product name.
4. `self.price = price` — stores the price.
5. `def apply_discount(self, percent):` — defines the discount behavior.
6. `if 0 <= percent <= 100:` — validates the percentage.
7. `self.price -= self.price * percent / 100` — updates the price using the discount.
8. `product = Product("Laptop", 100000)` — creates the object.
9. `product.apply_discount(20)` — applies a 20% discount, leaving 80000.0.
10. `print(product.price)` — prints 80000.0.

## Question 100

**Question:** Create one small real-world example using this concept.

In [100]:
class Product:
    def __init__(self, name, price):
        self.name = name
        self.price = price

    def apply_discount(self, percent):
        if 0 <= percent <= 100:
            self.price -= self.price * percent / 100

products = [
    Product("Laptop", 100000),
    Product("Phone", 60000)
]

for product in products:
    product.apply_discount(10)
    print(product.name, product.price)

Laptop 90000.0
Phone 54000.0


### Line-by-line explanation

1. `class Product:` — defines the reusable product blueprint.
2. `def __init__(self, name, price):` — initializes product objects.
3. `self.name = name` — stores each product’s individual name.
4. `self.price = price` — stores each product’s individual price.
5. `def apply_discount(self, percent):` — defines a reusable discount method.
6. `if 0 <= percent <= 100:` — validates the discount.
7. `self.price -= self.price * percent / 100` — updates the product price.
8. `products = [...]` — creates multiple product objects.
9. `for product in products:` — processes each product using the same code.
10. `product.apply_discount(10)` — applies the same business rule to every product.
11. `print(product.name, product.price)` — prints each product’s updated data.

---

## 👨‍💻 About This Course

This **free Python course** was designed and developed by **Muhammad Saad** as a contribution toward making programming education more accessible. I created it to share what I’ve learned with students and beginners who want to learn Python but may not have access to paid courses.

The goal is simple: **to make learning Python easier, accessible, and beginner-friendly.** If this course helps someone learn a new skill or take their first step into programming, then this initiative has achieved its purpose.

If you found this class helpful, feel free to share it with someone who might benefit from it.
